source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -138,10 +146,39 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void ProxyFilter::onLoadDnsCacheComplete(const Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
-  ENVOY_STREAM_LOG(debug, "load DNS cache complete, continuing", *decoder_callbacks_);
+void ProxyFilter::addHostAddressToFilterState(
+    const Network::Address::InstanceConstSharedPtr& address) {
+  ENVOY_STREAM_LOG(trace, "Adding resolved host {} to filter state", *decoder_callbacks_,
+                   address->asString());
+  const Envoy::StreamInfo::FilterStateSharedPtr& filter_state =
+      decoder_callbacks_->streamInfo().filterState();
+
+  if (!filter_state->hasData<StreamInfo::AddressSetAccessor>(
+          StreamInfo::KEY_DYNAMIC_PROXY_UPSTREAM_ADDR)) {
+    auto address_set = std::make_unique<StreamInfo::AddressSetAccessorImpl>();
+    address_set->add(address);
+
+    filter_state->setData(StreamInfo::KEY_DYNAMIC_PROXY_UPSTREAM_ADDR, std::move(address_set),
+                          StreamInfo::FilterState::StateType::ReadOnly,
+                          StreamInfo::FilterState::LifeSpan::Request);
+
+  } else {
+    StreamInfo::AddressSetAccessor& address_set =
+        filter_state->getDataMutable<StreamInfo::AddressSetAccessor>(
+            StreamInfo::KEY_DYNAMIC_PROXY_UPSTREAM_ADDR);
+    address_set.add(address);
```

Comment:
is the right thing to do here to add the address, or replace the address?
I'd think if we did recreateStream (internal redirect) we'd want to remove old addresses rather than append to them, to make sure we only had the addresses we wanted to use

Reply:
Do you think we should add a `clear` API so that the caller can then call `clear` and `add`? Or is it better to replace the set with a single address?

Reply:
yeah I think in the long run we're going to want to support multiple addresses, but also handle recreateStream, so either overwrite an existing set, or clear it and add new data.
(if that doesn't make sense please ping me on slack)